vmdb/app/controllers/miq_request_controller.rb

```diff
@@ -426,112 +426,54 @@ def get_request_tab_type
-            @or_hash["or"] = Array.new
-            cond_hash = Hash.new
-            cond_hash["="] = {"value"=>k.to_s,"field"=>"MiqRequest-resource_type"}
-            @or_hash["or"].push(cond_hash)
-          elsif i == request_types.length-1
-            cond_hash = Hash.new
-            cond_hash["="] = {"value"=>k.to_s,"field"=>"MiqRequest-resource_type"}
-            @or_hash["or"].push(cond_hash)
-          else
-            cond_hash = Hash.new
-            cond_hash["="] = {"value"=>k.to_s,"field"=>"MiqRequest-resource_type"}
-            @or_hash["or"].push(cond_hash)
-          end
-        end
-      end
+    if (a_s = opts[:applied_states].presence)
+      cond.push("or" => a_s.collect { |s| {"=" => {"value" => s, "field" => "MiqRequest-approval_state"}} })
     end
-    cond.push(@or_hash)
+
+    cond.push("or" => request_types_for_model.keys.collect { |k| {"=" => {"value" => k.to_s, "field" => "MiqRequest-resource_type"}} })
 
     if opts[:type_choice] && opts[:type_choice] != "all"  # Add request_type filter, if selected
-      cond_hash = Hash.new
-      cond_hash["="] = {"value"=>opts[:type_choice],"field"=>"MiqRequest-request_type"}
-      cond.push(cond_hash)
-    end
-
-    if opts[:reason_text]  && opts[:reason_text] != ""
-      cond_hash = Hash.new
-      if opts[:reason_text].starts_with?("*") && opts[:reason_text].ends_with?("*")   # Replace beginning/ending * chars with % for SQL
-        hash_key = "INCLUDES"
-        reason_text = opts[:reason_text][1..-2]
-      elsif opts[:reason_text].starts_with?("*")
-        hash_key = "ENDS WITH"
-        reason_text = opts[:reason_text][1..-1]
-      elsif opts[:reason_text].ends_with?("*")
-        hash_key = "STARTS WITH"
-        reason_text = opts[:reason_text][0..-2]
-      else
-        hash_key = "INCLUDES"
-        reason_text = opts[:reason_text]
-      end
+      cond.push("=" => {"value" => opts[:type_choice], "field" => "MiqRequest-request_type"})
+    end
 
-      cond_hash["#{hash_key}"] = {"value"=>reason_text,"field"=>"MiqRequest-reason"}
-      cond.push(cond_hash)
+    if opts[:reason_text].present?
+      cond.push(prov_condition_reason_text_expression_key => {"value" => prov_condition_reason_text_sanitized, "field" => "MiqRequest-reason"})
     end
 
-    condition = Hash.new
-    condition["and"] = Array.new
-    cond.each do |c|
-      condition["and"].push(c)
+    MiqExpression.new("and" => cond)
+  end
+
+  def request_types_for_model
+    MiqRequest::MODEL_REQUEST_TYPES[model_request_type_from_layout]
+  end
+
+  def model_request_type_from_layout
+    case @layout
+    when "miq_request_ae"   then :AutomationRequest
+    when "miq_request_host" then :Host
+    else                         :Vm
     end
-    return  MiqExpression.new(condition)
+  end
+
+  def prov_condition_reason_text_sanitized
+    opts[:reason_text].gsub(/\A\*/, "").chomp("*")  # Remove leading and/or trailing "*"
```

Comment:
That looks *EXACTLY* like the one I sent @brandondunne to begin with :stuck_out_tongue: , except I had sub, and there no need for the block:

```ruby
s.sub!(/\A\*?(.+?)\*?\z/, "\1")
```